### ISSUE TITLE
[Feature request] Keyboard catchall modes

### ISSUE THREAD

**Issue by buhl** [i3/i3#2448]
I think it would be a nice feature to be able to create modes where keypresses not matching any bindsym defined within would be caught and not passed on to containers.

Right now I can continue to write to my terminal window in resize mode. I would like to disallow this.

I would like to be able to write something like
```
mode --catch-all "resize" {
        bindsym j resize shrink width 10 px or 10 ppt
        bindsym k resize grow height 10 px or 10 ppt
        bindsym l resize shrink height 10 px or 10 ppt
        bindsym semicolon resize grow width 10 px or 10 ppt
        bindsym Escape mode "default"
}
```
in my config.

I am willing to try to make a patch to this on my own if it's something the project would like?
I have little experience with i3 and C but i like your documentation and I think I may, with a little help, be able to add this feature.

**Comment by stapelberg** [comment]
https://github.com/i3/i3/issues/249 is related (but the original reporter never finished working on the patch).

Before we get to the implementation, though, let me ask you what the motivation behind your use-case is. Why is it bad that you can keep writing into your terminal while in resize mode?

**Comment by buhl** [comment]
I would like my usage of i3 to be more modal with simpler keybindings.
For example I would like to define a "normal" mode where moving between containers  would have very simple bindings h,j,k,l or between or workspaces 1, 2, 3 ... I would rather like to move between modes than appending a modifier key to bindings.

I know this could still be done without the catchall feature. But with this simpler keybindings your are also more likely to send keys to a container that can act on these. I think it's more a principle matter than it is useful. But I like to know that when I am in a mode I cant unintentionally affect another mode.

Unlike the ticket you liked to I would like to indicate an error if you press a key that has not been bound, rather than escape to default mode.

I hope I was able to explain my self. And I totally understand if this is not something you would like to see in your project, let me know what you think.

**Comment by stapelberg** [comment]
Thanks for the additional context.

I think the idea of one-letter keys combined with modes (much like in Vi) is interesting, and in fact in i3 v3.x we had one-letter keys for all commands. Back then, support for actual modes wasn’t implemented yet (and visual cues about the current mode came even later). Actual user demand for any sort of combination/chaining of one-letter keys was low, so we ended up moving to more verbose commands.

Given that we’ve had limited success, albeit in a slightly different form, I think before changing i3 we’d like to see a proof of concept first so that we can judge the look&feel, see how users like it, and think some more about the details. This could be an external tool, which seems doable to me, or your own branch of i3.

**Comment by buhl** [comment]
Sure, my thought was always to work in a fork of i3. I am not proficient C coder and still have limited experience with i3's source code. So it will probably take some time to get something worthwhile adopting into the code base. I just wanted to hear if there was any interest for this before I started spending time on it.
As for which solution I should go for, I am still a little in doubt. The --catch-all flag solution seems more intuitive. but a catch all binding would make it more flexible as to what a user might want to happen when a key, not bound, is pressed.
Do you by any chance still have the patch referred to in #249 ? The link is dead.

**Comment by stapelberg** [comment]
I don’t have that patch anymore. http://cr.i3wm.org/patch/531/ might be helpful, but the keyboard-handling code changed significantly in the meanwhile.

**Comment by buhl** [comment]
Thanks, I will give it a shot :) We can close this issue if you like?

**Comment by pstray** [comment]
A method of ignoring any other key than the ones bound in a mode, or exiting the mode if any nonbound key is pressed would be quite useful.  "bindkey Any" or something other?

**Comment by buhl** [comment]
I finally got some time to work on this.

I first thought that I would define a new configuration keyword ala binddef. and then have a mode->catchallbind. But this seems like a big intrusion on the code. 

So instead, I applied the patch from http://cr.i3wm.org/patch/531/ manually against i3-4.12 and have this working locally now.

As described in the the link, it has too be the last definition in your mode as bindings are cycled in the order they are being defined in the configuration.

This is the diff I am currently using

```diff
diff --git a/src/bindings.c b/src/bindings.c
index eec821b..08e705d 100644
--- a/src/bindings.c
+++ b/src/bindings.c
@@ -61,7 +61,10 @@ Binding *configure_binding(const char *bindtype, const char *modifiers, const ch
     new_binding->release = (release != NULL ? B_UPON_KEYRELEASE : B_UPON_KEYPRESS);
     new_binding->border = (border != NULL);
     new_binding->whole_window = (whole_window != NULL);
-    if (strcmp(bindtype, "bindsym") == 0) {
+    if (strcasecmp(input_code, "any") == 0) {
+        new_binding->input_type = B_KEYBOARD;
+        new_binding->keycode = XCB_GRAB_ANY;
+    } else if (strcmp(bindtype, "bindsym") == 0) {
         new_binding->input_type = (strncasecmp(input_code, "button", (sizeof("button") - 1)) == 0
                                        ? B_MOUSE
                                        : B_KEYBOARD);
@@ -151,7 +154,7 @@ void grab_all_keys(xcb_connection_t *conn) {
             continue;
 
         /* The easy case: the user specified a keycode directly. */
-        if (bind->keycode > 0) {
+        if (!bind->symbol || bind->keycode > 0) {
             grab_keycode_for_binding(conn, bind, bind->keycode);
             continue;
         }
@@ -263,7 +266,7 @@ static Binding *get_binding(i3_event_state_mask_t state_filtered, bool is_releas
                 continue;
 
             /* This case is easier: The user specified a keycode */
-            if (bind->keycode != input_code)
+            if (bind->keycode != input_code && bind->keycode != XCB_GRAB_ANY)
                 continue;
         }
 
@@ -468,7 +471,7 @@ void translate_keysyms(void) {
             continue;
         }
 
-        if (bind->keycode > 0)
+        if (!bind->symbol)
             continue;
 
         /* We need to translate the symbol to a keycode */
```

I don't know if this is the best solution? If you guys think this is viable I will try to write some test cases and make a pull request.

comments are welcome :)

**Comment by fgrsnau** [comment]
I would love to have this feature in i3. I think it makes using modes much more convenient. Sometimes I’m stuck in a mode without recognizing and only half of the keys have the desired effect. I use the above patch with either `bind any mode "default"` or `bind any mode nop` which I find very useful.

**Comment by fgrsnau** [comment]
I just want to add: The patch works almost as expected. After restructuring my config I observed that any modifier key also triggers the `any` binding. So binding `$modifier+key` is not possible.

Ignoring all modifier keys for this case in the `grab_all_keys` function is most likely not very complicated.

**Comment by buhl** [comment]
@fgrsnau sorry for the delay. I am not sure I understand the problem. Could you give me an example configuration where the problem exists and a descriptions on how to reproduce it?

**Comment by foonathan** [comment]
Is this still being worked on?

**Comment by buhl** [comment]
No, I haven't touched this in some time. I don't even know if the patch still applies cleanly.
